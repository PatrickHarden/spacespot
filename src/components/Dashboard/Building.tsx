import React, { useState } from 'react'
import { useIntl } from 'react-intl'
import { useDispatch, useSelector } from 'react-redux'
import styled from 'styled-components'
import { Link } from 'react-router-dom'
import { TranslatePropSpace } from 'services/space'
import actions from 'services/onboarding/actions'
import selectors from 'services/dashboard/selectors'
import PublishedSpace from '../OnBoarding/PublishedSpace'
import Trash from 'components/icons/Trash'
import ConfirmDialog from 'components/common/ConfirmDialog'
import Heading2 from 'components/common/Heading2'
import OutlinedButton from 'components/common/OutlinedButton'
import HTMLContent from 'components/common/HTMLContent'
import DOMPurify from 'dompurify'

const Container = styled.div<{
  deleted: boolean
}>`
  max-width: 1042px;
  margin-bottom: 60px;
  ${props =>
    props.deleted &&
    `
  opacity: 0.2;
  display: none;
  `}
`

const Row = styled.div`
  display: flex;
  justify-content: space-between;
`

const Title = styled(Heading2)`
  margin-bottom: 0.3em;
`

const Address = styled.div`
  width: 100%;
  color: #828286;
  font-size: 16px;
  font-weight: 400;
  line-height: 21px;
`
const Edit = styled(Link)`
  display: block;
  margin-top: 45px;
  color: #4fbbd8;
  font-size: 16px;
  font-weight: 400;
  line-height: 21px;
  text-decoration: none;
`

const DeleteButton = styled(OutlinedButton)`
  color: #404042;
  background-color: #ffffff;
  border: 1px solid #000000;
  box-sizing: border-box;
  display: flex;
  align-items: center;

  svg {
    margin-right: 6px;
    padding-top: 2px;
  }
`

const shorten = (name?: string) => {
  return !name || name.length < 100 ? name : name.substring(0, 96) + '...'
}

const Building = (props: {
  building: TranslatePropSpace
  spaces: Array<TranslatePropSpace>
}) => {
  const { formatMessage } = useIntl()
  const dispatch = useDispatch()
  const [showPopup, setShowPopup] = useState(false)
  const t = (s: string) => formatMessage({ id: s })
  const pendingDelete = useSelector(selectors.pendingDelete)
  const { building, spaces } = props
  const isPendingDelete = pendingDelete.includes(building.key)

  const deleteBuilding = () => {
    dispatch(actions.deleteBuilding(building.key))
    setShowPopup(false)
  }
  const location = DOMPurify.sanitize(shorten(building.location) as string)

  return (
    <Container deleted={isPendingDelete}>
      <Row>
        <div>
          <Title>{building.address}</Title>
          <Address>
            <HTMLContent html={location} />
          </Address>
        </div>
        <div>
          <Edit to={`/onboarding/edit/building/${building.key}`}>
            {t('ONBOARDING_EDIT_BUILDING')}
          </Edit>
        </div>
      </Row>
      {(spaces && spaces.length) > 0 ? (
        spaces.map((space: TranslatePropSpace) => (
          <PublishedSpace key={space.key} space={space} showControls={true} />
        ))
      ) : (
        <>
          <p>{t('DASHBOARD_NO_SPACES')}</p>
          <DeleteButton
            data-testid="delete-building"
            onClick={() => setShowPopup(true)}>
            <span>
              <Trash color="#000" />
            </span>
            {t('ONBOARDING_DELETE_BUILDING')}
          </DeleteButton>
        </>
      )}
      {showPopup && (
        <ConfirmDialog
          open={showPopup}
          setOpen={setShowPopup}
          onClick={deleteBuilding}
          title={t('ONBOARDING_REMOVE_BUILDING_POPUP_TITLE')}
          cancelText={t('ONBOARDING_REMOVE_BUILDING_POPUP_KO')}
          okText={t('ONBOARDING_REMOVE_BUILDING_POPUP_OK')}
        />
      )}
    </Container>
  )
}
export default Building
