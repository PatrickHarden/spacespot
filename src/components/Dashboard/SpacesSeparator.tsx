import React from 'react'
import { useIntl } from 'react-intl'
import styled from 'styled-components'

const Container = styled.div`
  width: 100%;
  text-align: left;
  position: relative;
`

const SpacesCount = styled.span`
  line-height: 1.5em;
  color: black;
  font-size: 16px;
  font-weight: 400;
  background-color: white;
  padding: 0 25px;
  margin-left: 20px;
`
const Line = styled.div`
  :after {
    content: ' ';
    position: absolute;
    top: 60%;
    left: 50%;
    width: 100%;
    border: 1px solid #979797;
    border-top: 0;
    border-radius: 4px;
    box-shadow: inset 0 1px 1px rgba(0, 0, 0, 0.05);
    transform: translateX(-50%);
    z-index: -2;
  }
`

const SpacesSeparator = (props: { num: number }) => {
  const { formatMessage } = useIntl()
  const tn = (s: string, value: number) =>
    formatMessage({ id: s }, { number: value })
  const { num } = props
  return (
    <>
      <Container>
        <Line>
          <SpacesCount>{tn('DASHBOARD_NUM_SPACES', num)}</SpacesCount>
        </Line>
      </Container>
    </>
  )
}
export default SpacesSeparator
