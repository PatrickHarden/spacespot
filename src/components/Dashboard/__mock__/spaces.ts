export default [
  {
    photo:
      'https://uatlistingssearchcbreeuw.blob.core.windows.net/fileassets/SS_SPACE_344/23c501c9/51-53%20St%20Martins%20Lane%201%2003.2019_Photo_1_small.jpg',
    key: 'SS_SPACE_344',
    title: 'SS_SPACE_344',
    street: 'Bryggegata 25, 0250 Oslo, Noruega',
    city: 'Oslo',
    postCode: '0250',
    description: '',
    latLng: { lat: 59.91078, lng: 10.726055 },
    aspects: undefined,
    rent: '',
    size: 0,
    floor: 0,
    availability: new Date('2020-02-02'),
    type: 'FLEX',
    hotDesks: 0,
    fixedDesks: 0,
    sevicedOffices: 0,
  },
  {
    photo:
      'https://uatlistingssearchcbreeuw.blob.core.windows.net/fileassets/SS_SPACE_343/23c501c9/51-53%20St%20Martins%20Lane%201%2003.2019_Photo_1_small.jpg',
    key: 'SS_SPACE_343',
    title: 'space name ',
    street: 'Bryggegata 25, 0250 Oslo, Noruega',
    city: 'Oslo',
    postCode: '0250',
    description: '',
    latLng: { lat: 59.91078, lng: 10.726055 },
    aspects: undefined,
    rent: '8000000kr',
    size: 1000,
    floor: 0,
    availability: new Date('2020-02-02'),
    type: 'FIXED',
    hotDesks: 0,
    fixedDesks: 0,
    sevicedOffices: 0,
  },
]
