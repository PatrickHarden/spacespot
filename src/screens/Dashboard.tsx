import React, { useEffect } from 'react'
import { useDispatch, useSelector } from 'react-redux'

import actions from 'services/dashboard/actions'
import selectors from 'services/dashboard/selectors'
import DashboardMain from 'components/Dashboard/Dashboard'
import OnBoardingLayout from 'components/OnBoardingLayout'
import Spinner from 'components/icons/Spinner'
import styled from 'styled-components'

const SpinnerC = styled((props: { className?: string }) => (
  <div className={props.className}>
    <Spinner />
  </div>
))`
  text-align: center;
  margin-top: 40px;
`

const Dashboard = () => {
  const dispatch = useDispatch()
  const initOK = useSelector(selectors.isDashboardInitOK)
  const loading = useSelector(selectors.isDashboardLoading)
  useEffect(() => {
    if (!initOK) {
      dispatch(actions.dashboardInit())
    }
  }, [dispatch, initOK])
  return (
    <OnBoardingLayout>
      {!initOK || loading ? <SpinnerC /> : <DashboardMain />}
    </OnBoardingLayout>
  )
}
export default Dashboard
