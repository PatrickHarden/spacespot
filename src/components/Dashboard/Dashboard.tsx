import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { useIntl } from 'react-intl'
import styled from 'styled-components'
import device from 'services/global/device'
import selectors from 'services/dashboard/selectors'
import actions from 'services/onboarding/actions'
import { TranslatePropSpace } from 'services/space'
import SpacesSeparator from './SpacesSeparator'
import Building from './Building'
import Heading1 from 'components/common/Heading1'
import PlusIcon from 'components/icons/Plus'
import FilledButton from 'components/common/FilledButton'
import Colors from 'assets/Colors'
import history from 'browserhistory'

const Title = styled(Heading1)``
const Container = styled.div`
  margin: 0 auto 40px auto;
  max-width: 1042px;
  width: 90vw;
`
const Row = styled.div`
  display: flex;
  align-items: center;
  justify-content: space-between;
  margin-top: 25px;
  @media ${device.mobile} {
    margin: 15px 0;
    flex-wrap: wrap;
  }
`

const AddButton = styled(FilledButton)`
  float: right;
  color: white;
  background: ${Colors.main.headerBg};
  cursor: pointer;
  span {
    margin-right: 10px;
  }
`

const Dashboard = () => {
  const dispatch = useDispatch()
  const spaces = useSelector(selectors.selectDashboardSpaces)
  const buildings = useSelector(selectors.selectDashboardBuildings)

  const { formatMessage } = useIntl()
  const t = (s: string) => formatMessage({ id: s })

  const countSpaces = spaces
    ? Object.keys(spaces).reduce(
        (numberAcumulator: number, key: string) =>
          numberAcumulator + spaces[key].length,
        0,
      )
    : 0
  return (
    <Container>
      <Row>
        <Title>{t('DASHBOARD_TITLE')}</Title>
        <AddButton
          data-testid="add-space"
          onClick={() => {
            dispatch(actions.resetSpaces())
            history.push(`/onboarding/`)
          }}>
          <span>
            <PlusIcon size="14px" color="white" />
          </span>
          {t('DASHBOARD_ADD_SPACE')}
        </AddButton>
      </Row>
      <SpacesSeparator num={countSpaces} />
      {buildings
        ? buildings.map((building: TranslatePropSpace) => (
            <Building
              key={building.key}
              building={building}
              spaces={spaces[building.key]}
            />
          ))
        : null}
    </Container>
  )
}
export default Dashboard
