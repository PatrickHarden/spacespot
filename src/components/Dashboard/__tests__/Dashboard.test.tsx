import React from 'react'
import { mount } from 'enzyme'
import { mountToJson } from 'enzyme-to-json'
import { Router } from 'react-router-dom'
import { createBrowserHistory } from 'history'
import { createMockStore } from 'redux-test-utils'
import { Provider } from 'react-redux'

import Dashboard from '../Dashboard'
import buildings from '../__mock__/buildingBackend'
import spaces from '../__mock__/spaceBackend'

jest.mock('react-intl', () => ({
  ...jest.requireActual('react-intl'),
  useIntl: () => ({ formatMessage: (s: any) => s.id }),
}))

it('renders without crashing', async () => {
  const testState = {
    space: {
      dashboard: {
        spaces,
        buildings,
      },
    },
  }
  const store: any = createMockStore(testState)
  const wrapper = mount(
    <Provider store={store}>
      <Router history={createBrowserHistory()}>
        <Dashboard />
      </Router>
    </Provider>,
  )
  expect(mountToJson(wrapper)).toMatchSnapshot()
})
