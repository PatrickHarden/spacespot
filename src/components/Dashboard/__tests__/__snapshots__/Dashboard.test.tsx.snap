// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getAction": [Function],
      "getActions": [Function],
      "getState": [Function],
      "isActionDispatched": [Function],
      "isActionTypeDispatched": [Function],
      "subscribe": [Function],
    }
  }
>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Dashboard>
      <styled.div>
        <StyledComponent
          forwardedComponent={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "attrs": Array [],
              "componentStyle": ComponentStyle {
                "componentId": "sc-jWBwVP",
                "isStatic": false,
                "lastClassName": "ilfUki",
                "rules": Array [
                  "
  margin: 0 auto 40px auto;
  max-width: 1042px;
  width: 90vw;
",
                ],
              },
              "displayName": "styled.div",
              "foldedComponentIds": Array [],
              "render": [Function],
              "styledComponentId": "sc-jWBwVP",
              "target": "div",
              "toString": [Function],
              "warnTooManyClasses": [Function],
              "withComponent": [Function],
            }
          }
          forwardedRef={null}
        >
          <div
            className="sc-jWBwVP ilfUki"
          >
            <styled.div>
              <StyledComponent
                forwardedComponent={
                  Object {
                    "$$typeof": Symbol(react.forward_ref),
                    "attrs": Array [],
                    "componentStyle": ComponentStyle {
                      "componentId": "sc-brqgnP",
                      "isStatic": false,
                      "lastClassName": "DZhsl",
                      "rules": Array [
                        "
  display: flex;
  align-items: center;
  justify-content: space-between;
  margin-top: 25px;
",
                      ],
                    },
                    "displayName": "styled.div",
                    "foldedComponentIds": Array [],
                    "render": [Function],
                    "styledComponentId": "sc-brqgnP",
                    "target": "div",
                    "toString": [Function],
                    "warnTooManyClasses": [Function],
                    "withComponent": [Function],
                  }
                }
                forwardedRef={null}
              >
                <div
                  className="sc-brqgnP DZhsl"
                >
                  <Styled(styled.h1)>
                    <StyledComponent
                      forwardedComponent={
                        Object {
                          "$$typeof": Symbol(react.forward_ref),
                          "attrs": Array [],
                          "componentStyle": ComponentStyle {
                            "componentId": "sc-cvbbAY",
                            "isStatic": false,
                            "lastClassName": "kUipAf",
                            "rules": Array [
                              "
  color: #404042;
  font-size: 40px;
  font-weight: 200;
",
                            ],
                          },
                          "displayName": "Styled(styled.h1)",
                          "foldedComponentIds": Array [
                            "sc-eHgmQL",
                          ],
                          "render": [Function],
                          "styledComponentId": "sc-cvbbAY",
                          "target": "h1",
                          "toString": [Function],
                          "warnTooManyClasses": [Function],
                          "withComponent": [Function],
                        }
                      }
                      forwardedRef={null}
                    >
                      <h1
                        className="sc-eHgmQL sc-cvbbAY kUipAf"
                      >
                        DASHBOARD_TITLE
                      </h1>
                    </StyledComponent>
                  </Styled(styled.h1)>
                  <Styled(styled.button)
                    data-testid="add-space"
                    onClick={[Function]}
                  >
                    <StyledComponent
                      data-testid="add-space"
                      forwardedComponent={
                        Object {
                          "$$typeof": Symbol(react.forward_ref),
                          "attrs": Array [],
                          "componentStyle": ComponentStyle {
                            "componentId": "sc-cMljjf",
                            "isStatic": false,
                            "lastClassName": "gLZRCD",
                            "rules": Array [
                              "
  font-size: 14px;
  font-weight: 550;
  line-height: 18px;
  border: none;
  background-color: #6cb9d5;
  color: #ffffff;
  border-radius: 2px;
  cursor: pointer;
  text-transform: uppercase;
  height: 37px;
  padding: 0 20px;
",
                              "
  float: right;
  color: white;
  background: ",
                              "#404042",
                              ";
  cursor: pointer;
  span {
    margin-right: 10px;
  }
",
                            ],
                          },
                          "displayName": "Styled(styled.button)",
                          "foldedComponentIds": Array [
                            "sc-ifAKCX",
                          ],
                          "render": [Function],
                          "styledComponentId": "sc-cMljjf",
                          "target": "button",
                          "toString": [Function],
                          "warnTooManyClasses": [Function],
                          "withComponent": [Function],
                        }
                      }
                      forwardedRef={null}
                      onClick={[Function]}
                    >
                      <button
                        className="sc-ifAKCX sc-cMljjf gLZRCD"
                        data-testid="add-space"
                        onClick={[Function]}
                      >
                        <span>
                          <Plus
                            color="white"
                            size="14px"
                          >
                            <svg
                              color="white"
                              height="14px"
                              size="14px"
                              viewBox="0 0 18 17"
                              width="14px"
                            >
                              <g
                                fill="none"
                                fillRule="evenodd"
                                stroke="white"
                                strokeLinecap="round"
                                strokeWidth={2}
                              >
                                <path
                                  d="M9 1.6v13.8M1.64 8.5h14.72"
                                />
                              </g>
                            </svg>
                          </Plus>
                        </span>
                        DASHBOARD_ADD_SPACE
                      </button>
                    </StyledComponent>
                  </Styled(styled.button)>
                </div>
              </StyledComponent>
            </styled.div>
            <SpacesSeparator
              num={0}
            >
              <styled.div>
                <StyledComponent
                  forwardedComponent={
                    Object {
                      "$$typeof": Symbol(react.forward_ref),
                      "attrs": Array [],
                      "componentStyle": ComponentStyle {
                        "componentId": "sc-bdVaJa",
                        "isStatic": false,
                        "lastClassName": "eUnCtI",
                        "rules": Array [
                          "
  width: 100%;
  text-align: left;
  position: relative;
",
                        ],
                      },
                      "displayName": "styled.div",
                      "foldedComponentIds": Array [],
                      "render": [Function],
                      "styledComponentId": "sc-bdVaJa",
                      "target": "div",
                      "toString": [Function],
                      "warnTooManyClasses": [Function],
                      "withComponent": [Function],
                    }
                  }
                  forwardedRef={null}
                >
                  <div
                    className="sc-bdVaJa eUnCtI"
                  >
                    <styled.div>
                      <StyledComponent
                        forwardedComponent={
                          Object {
                            "$$typeof": Symbol(react.forward_ref),
                            "attrs": Array [],
                            "componentStyle": ComponentStyle {
                              "componentId": "sc-htpNat",
                              "isStatic": false,
                              "lastClassName": "xOUHL",
                              "rules": Array [
                                "
  :after {
    content: ' ';
    position: absolute;
    top: 60%;
    left: 50%;
    width: 100%;
    border: 1px solid #979797;
    border-top: 0;
    border-radius: 4px;
    box-shadow: inset 0 1px 1px rgba(0, 0, 0, 0.05);
    transform: translateX(-50%);
    z-index: -2;
  }
",
                              ],
                            },
                            "displayName": "styled.div",
                            "foldedComponentIds": Array [],
                            "render": [Function],
                            "styledComponentId": "sc-htpNat",
                            "target": "div",
                            "toString": [Function],
                            "warnTooManyClasses": [Function],
                            "withComponent": [Function],
                          }
                        }
                        forwardedRef={null}
                      >
                        <div
                          className="sc-htpNat xOUHL"
                        >
                          <styled.span>
                            <StyledComponent
                              forwardedComponent={
                                Object {
                                  "$$typeof": Symbol(react.forward_ref),
                                  "attrs": Array [],
                                  "componentStyle": ComponentStyle {
                                    "componentId": "sc-bwzfXH",
                                    "isStatic": false,
                                    "lastClassName": "dfhtkd",
                                    "rules": Array [
                                      "
  line-height: 1.5em;
  color: black;
  font-size: 16px;
  font-weight: 400;
  background-color: white;
  padding: 0 25px;
  margin-left: 20px;
",
                                    ],
                                  },
                                  "displayName": "styled.span",
                                  "foldedComponentIds": Array [],
                                  "render": [Function],
                                  "styledComponentId": "sc-bwzfXH",
                                  "target": "span",
                                  "toString": [Function],
                                  "warnTooManyClasses": [Function],
                                  "withComponent": [Function],
                                }
                              }
                              forwardedRef={null}
                            >
                              <span
                                className="sc-bwzfXH dfhtkd"
                              >
                                DASHBOARD_NUM_SPACES
                              </span>
                            </StyledComponent>
                          </styled.span>
                        </div>
                      </StyledComponent>
                    </styled.div>
                  </div>
                </StyledComponent>
              </styled.div>
            </SpacesSeparator>
          </div>
        </StyledComponent>
      </styled.div>
    </Dashboard>
  </Router>
</Provider>
`;
